neqo-transport/src/fc.rs

```diff
@@ -161,9 +164,131 @@ impl SenderFlowControl<StreamId> {
     }
 }
 
+#[derive(Debug)]
+pub struct ReceiverFlowControl<T>
+where
+    T: Debug + Sized,
+{
+    /// The thing that we're counting for.
+    subject: T,
+    /// Max amount of outstanding data
+    max_wnd: u64,
+    // Last max data sent.
+    max_data: u64,
+    // Retired bytes.
+    retired: u64,
+    frame_pending: bool,
+}
+
+impl<T> ReceiverFlowControl<T>
+where
+    T: Debug + Sized,
+{
+    /// Make a new instance with the initial value and subject.
+    pub fn new(subject: T, max_bytes: u64) -> Self {
+        Self {
+            subject,
+            max_wnd: max_bytes,
+            max_data: max_bytes,
+            retired: 0,
+            frame_pending: false,
+        }
+    }
+
+    /// Check if received data exceeds the allowed flow control limit.
+    pub fn check_allowed(&self, new_end: u64) -> bool {
+        new_end <= self.max_data
+    }
+
+    /// Some data has been read, retired them and maybe send flow control
+    /// update.
+    pub fn retired(&mut self, retired: u64) {
+        if retired <= self.retired {
+            return;
+        }
```

Comment:
👍 I was going to suggest this be an assertion, but this is better.  This way, we can use this for stream credit, which might not always retire streams in order.